### DISCUSSION TITLE
API et retour

### DISCUSSION THREAD

**Discussion by PhilippeLRETQC** [MTESSDev/FRW#16] [Annonces]
Bonjour,

Les différentes routes retournent vraiment uniquement des codes 200 même en cas d'erreur?

Bonne journée

**Comment by anisite** [comment]
Bonjour, 

il est possible d'avoir des 400 des 500 des même des 401 dans certains API, parfois l'erreur est caché par sécurité en 200 pour ne pas donner d'indices mais votre question vient du fait que le swagger ne les documente pas est-ce bien cela ?

**Comment by PhilippeLRETQC** [comment]
C'est exact. Comme ils ne sont pas documenté, il est plus difficile de prévoir les cas à supporter.

**Comment by anisite** [comment]
Ok, toutes mes excuses, je n'ai jamais documenter toutes les erreurs possible sur mes API, je pourrais corriger celle-ci pour plus tard, le swagger est généré à partir des Data annotations du code, nous n'avons pas inscrit les codes et leur causes.

Par contre en gros il ne sera pas possible de lister tous les codes d'erreurs pouvant survenir (certains pouvant provenir d'un reverse proxy entre autre) sachez que si le code n'est pas 200 vous devriez vous arrêter et consigner l'erreur, donc en général les code connus et lancés par l'application sont :

 - 200 : tout est beau
 - 400 : la route ou l'url est invalide/mal formée ou encore accès refusé (clée api invalide)
 - 404 : l'id du système autorisé, le type de formulaire ou autre partie de l'adresse est introuvable
 - 401 : Accès refusé (n'est retourné sur l'api "Document/Telecharger")
 - 500 : le serveur a planté

Certains API publics font exprès de retourner des codes 400 quand la clée API est invalide pour ne pas donner d'indices sur la validité des accès VS des id d'éléments ou autres, c'est ce qui avait été choisi dans ce cas-ci, mais il n'est pas trop tard pour qu'on switch ça à 401 pour respecter le REST.

Nous allons ajouter le code aussi pour émettre les types d'erreur connus. 

Merci

**Comment by IhsaneRQ** [comment]
Bonjour,

Pour faire la gestion des erreur dans notre API, nous avons besoin d'avoir la liste des codes erreurs que vous gérer ainsi leurs descriptions, lors des appels à l'API du MESS, j'ai vu votre discussion avec Philippe, mais j'aimerai savoir si vous les avez mis à jour ou documenté depuis ce temps la.

Merci d'avance et bonne fin de journée

**Comment by anisite** [comment]
Bonjour @IhsaneRQ 

J'ai remis à jour ce matin le swagger disponible ici : https://github.com/MTESSDev/FRW/tree/main/Swagger

J'ai ainsi ajouté les différents codes de retour HTTP des différents services.

Il n'y a pas plus de documentation que ça de prévue, sachez aussi que c'est sujet à changement, demain matin il pourrait y avoir d'autre codes HTTP ou certains 500 pourrait devenir des 400 ou vice-versa.

Si nous vous retournons autre chose qu'un code de succès vous devriez en tout temps le gérer comme une erreur, de la même façon générique si possible.

Dernier point, nous allons rendre l'interface Swagger disponible en IT pour les partenaires.

**Comment by IhsaneRQ** [comment]
Bonjour @anisite, 

Ok, parfait, merci beaucoup pour la rapidité de retour et pour ces détails

Bonne journée

**Comment by anisite** [comment]
Le lien vers le Swagger est maintenant disponible en IT 
![image](https://user-images.githubusercontent.com/3194288/236253109-ec628c64-6008-4128-8ccc-0e85a9ceb8d0.png)

Si une de nos réponse vous convient, SVP marquer la discussion comme répondu. Merci

**Comment by Hutson48** [comment]
Ceci est disponible dans notre documentation : https://github.com/MTESSDev/FRW/tree/main/Swagger
